### PR TITLE
Add Application insights to React App

### PR DESCRIPTION
**Pull request by mhassaninmsft**
1. Updates the Front end application to use Application insights using the MSFT react plugin. []
2. Updates instructions for creation of dev environments to make it safer and easier to update application insights `ikey` and signalR service backend.

### CODE REVIEW

File: RetrospectiveExtension.Backend/appsettings.json [hunk 1]
```diff
@@ -11,5 +11,8 @@
         "Default": "Warning"
       }
     }
+  },
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Backend/appsettings.json:14]: Adding Default placeholder in appsettings for the backend to make it for local debugging/testing to change this

File: RetrospectiveExtension.Frontend/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Development version to publish dev version of extension
+vss-extension-dev.json
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/.gitignore]: do we need this nested `.gitignore`? If so, the commented out one should be removed from the main one. But i'm not sure we need a separate one for the frontend.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -38,11 +38,20 @@ Clone the repository to your local machine from the Azure DevOps endpoint.
...
+   "AppInsightsInstrumentKey": "my_instrumentation_key" // put Instrumentation key here
+}
+```
+
 - Run `npm run build` to build the project. Refer to the 'scripts' section in 'package.json' for other commands.
 
 - To test your changes, you will need to publish a new extension under a new Azure DevOps publisher account. Refer to the [documentation](https://docs.microsoft.com/en-us/azure/devops/extend/publish/overview?view=vsts) on publishing extensions. You can publish it to any test Azure DevOps organization that you are an admin of (As a Microsoft employee, you can create a new test organization from your Azure DevOps profile page). Currently this is the only way to test the extension.
 
-- Update the 'vss-extension-dev.json' file with the new publisher that you setup. Also update the name and id fields.
+- Copy the file `vss-extensions-dev.json.template` into `vss-extensions-dev.json` file with the new publisher that you setup. Also update the name and id fields.
```
Comment (polatengin) [inline CONTRIBUTING.md]: `vss-extension-dev.json` (singular) file ignored in the `.gitignore` file, here we're stating the file as plural `(vss-extensions-dev.json)`

either of them should be changed I think.

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx [hunk 4]
```diff
@@ -24,7 +26,7 @@ export default class FeedbackCarousel extends React.Component<IFeedbackCarouselP
       .filter((columnItem) => !columnItem.feedbackItem.parentFeedbackItemId)
       .map((columnItem) => {
         const feedbackItemProps =
-          FeedbackColumn.createFeedbackItemProps(feedbackColumnProps, columnItem, true);
+        FeedbackColumnStatic.createFeedbackItemProps(feedbackColumnProps, columnItem, true);
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: I'd move `createFeedbackItemProps` static function to a utility class and use it, instead of exporting same class with two different approaches (withAITracking and with static)

it seems `createFeedbackItemProps` method is the only method we need here

File: RetrospectiveExtension.Frontend/config/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# ignore enviornment containing possibly sensitive information such as the backend/ instrumentation key
+environment.tsx
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/config/.gitignore]: Or this separate `.gitignore` - it makes more sense to me to just have the toplevel one and specify all the files that may have sensitive keys in them.

File: RetrospectiveExtension.Frontend/components/feedbackColumn.tsx [hunk 6]
```diff
@@ -300,3 +302,7 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
     return this.renderFeedbackColumn();
   }
 }
+
+export { FeedbackColumn as FeedbackColumnStatic}; // static functions are not preserved with `withAITracking`,
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackColumn.tsx]: can we move static methods to a utility class and not export the same class in a different approach?

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx [hunk 7]
```diff
@@ -934,7 +936,7 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
               }
               {this.state.searchedFeedbackItems
                 .map((columnItem) => {
-                  const feedbackItemProps = FeedbackColumn.createFeedbackItemProps(
+                  const feedbackItemProps = FeedbackColumnStatic.createFeedbackItemProps(
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: same with the previous comment

File: RetrospectiveExtension.Frontend/config/environment.tsx.template [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const environment = {
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/config/environment.tsx.template:1]: I appreciate this being template-able! 👏

File: RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx [hunk 9]
```diff
@@ -49,7 +51,7 @@ export default class FeedbackItemGroup extends React.Component<IFeedbackItemGrou
     // Bug 19016440: Edge drag and drop dataTransfer protocol is bugged
     // const droppedItemId = e.dataTransfer.getData('id');
     const droppedItemId = localStorageHelper.getIdValue();
-    FeedbackItem.handleDropFeedbackItemOnFeedbackItem(this.props.mainFeedbackItem, droppedItemId, this.props.mainFeedbackItem.id);
+    FeedbackItemStatic.handleDropFeedbackItemOnFeedbackItem(this.props.mainFeedbackItem, droppedItemId, this.props.mainFeedbackItem.id);
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx]: same with the previous comment

_this time for `handleDropFeedbackItemOnFeedbackItem` method_

File: RetrospectiveExtension.Frontend/index.tsx [hunk 10]
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 import { isHostedAzureDevOps } from './utilities/azureDevOpsContextHelper';
 import { getProjectId } from './utilities/servicesHelper';
 import './css/main.scss';
+import {  appInsights } from './utilities/external/telemetryClient';
 
 import FeedbackBoardContainer, { FeedbackBoardContainerProps } from './components/feedbackBoardContainer';
-// TODO (enpolat) : import { appInsightsClient, TelemetryEvents } from './utilities/appInsightsClient'
+// import { TelemetryClient } from './utilities/external/telemetryclient';
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/index.tsx]: We shouldn't be committing commented out code

File: RetrospectiveExtension.Frontend/index.tsx [hunk 11]
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
...
 
 initializeIcons();
-
-// TODO (enpolat) : appInsightsClient.trackEvent(TelemetryEvents.ExtensionLaunched);
+appInsights.trackEvent({name:"Startup",properties:{"version":"2"}});
+appInsights.trackMetric({name:"startup Time",average:5});
+// let client = TelemetryClient.getClient();
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/index.tsx]: Mindful of commented out code.

**Review by polatengin — APPROVED** [review]
I left a few comments, but so far it looks very good. if you can address those comments, it'd be even better. thanks 🥳👍

File: RetrospectiveExtension.Frontend/components/workflowStage.tsx [hunk 12]
```diff
@@ -1,6 +1,8 @@
 ﻿import classNames from 'classnames';
 import * as React from 'react';
 import { WorkflowPhase } from '../interfaces/workItem';
+import { withAITracking } from '@microsoft/applicationinsights-react-js';
+import { reactPlugin, appInsights } from '../utilities/external/telemetryClient';
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/components/workflowStage.tsx:5]: I'd assume this would fail if we're removing that elsewhere in the PR?

File: RetrospectiveExtension.Frontend/components/noFeedbackBoardsView.tsx [hunk 13]
```diff
@@ -24,3 +26,6 @@ export default class NoFeedbackBoardsView extends React.Component<NoFeedbackBoar
     )
   }
 }
+
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/components/noFeedbackBoardsView.tsx:29]: Probably only need one new line here

File: RetrospectiveExtension.Backend/appsettings.json [hunk 14]
```diff
@@ -11,5 +11,8 @@
...
+  "ApplicationInsights":{
```
Comment (mindlessroman) [inline RetrospectiveExtension.Backend/appsettings.json]: Just for uniformity's sake - probably could have spaces between the colons and the subsequent characters

File: RetrospectiveExtension.Frontend/components/workflowStage.tsx (hunk 12 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/workflowStage.tsx:5]: NEVERMIND.... I didn't see the capitalization change

File: RetrospectiveExtension.Frontend/components/selectorCombo.tsx [hunk 15]
```diff
@@ -300,3 +302,6 @@ export default class SelectorCombo<T> extends React.Component<ISelectorComboProp
     });
   }
 }
+
+export default  SelectorCombo;
+// export default withAITracking(reactPlugin,SelectorCombo); //mhassani: TODO: Figure out how to make App insights export generic classes
```
Comment (mindlessroman) [inline RetrospectiveExtension.Frontend/components/selectorCombo.tsx]: No lingering TODOs before merging. (mentioning here to keep track of it)

File: RetrospectiveExtension.Frontend/components/noFeedbackBoardsView.tsx (hunk 13 above)
Comment (mindlessroman) [reply RetrospectiveExtension.Frontend/components/noFeedbackBoardsView.tsx:29]: (and elsewhere)

File: .gitignore [hunk 16]
```diff
@@ -266,3 +266,5 @@ RetrospectiveExtension.FrontEnd/vss-extension.dev.json
 .env
 allowed_origins.json
 dev_certs.json
+
```
Comment (mindlessroman) [inline .gitignore]: We should include anything we provide templates for here.

**Review by mindlessroman — COMMENTED** [review]
I had a general question about the naming convention of "withAITracking" - would love to know more.

File: RetrospectiveExtension.Frontend/components/feedbackColumn.tsx (hunk 6 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackColumn.tsx]: makes sense

File: .gitignore (hunk 16 above)
Comment (phongcao) [reply .gitignore]: Similar to Hannah's concern, can we remove line 270 to make the file look cleaner?

File: RetrospectiveExtension.Frontend/components/actionItemDisplay.tsx [hunk 17]
```diff
@@ -316,3 +318,4 @@ export default class ActionItemDisplay extends React.Component<ActionItemDisplay
     );
   }
 }
+export default withAITracking(reactPlugin,ActionItemDisplay);
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/actionItemDisplay.tsx]: nit: Add a space after `,`. Please fix this issue in other places as well.

File: RetrospectiveExtension.Frontend/components/effectivenessMeasurementRow.tsx [hunk 18]
```diff
@@ -20,8 +22,7 @@ export interface EffectivenessMeasurementRowProps {
 export interface EffectivenessMeasurementRowState {
   selected: number;
 }
-
-export default class EffectivenessMeasurementRow extends React.Component<EffectivenessMeasurementRowProps, EffectivenessMeasurementRowState> {
+ class EffectivenessMeasurementRow extends React.Component<EffectivenessMeasurementRowProps, EffectivenessMeasurementRowState> {
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/effectivenessMeasurementRow.tsx:26]: nit: add empty line above. Same for other places.

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx [hunk 19]
```diff
@@ -77,7 +79,7 @@ export default class FeedbackCarousel extends React.Component<IFeedbackCarouselP
             {mainCardCount >= 2 &&
               // @ts-ignore TS2786
               <Slider {...settings}>
-                {React.Children.map(this.renderFeedbackCarouselItems(columnProps), (child: React.ReactElement<FeedbackItem>) => {
+                {React.Children.map(this.renderFeedbackCarouselItems(columnProps), (child: React.ReactElement<typeof FeedbackItem>) => {
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx:82]: Can you explain why we need to add `typeof` here?

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx [hunk 20]
```diff
@@ -977,3 +979,7 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
     );
   }
 }
+
+export {FeedbackItem as FeedbackItemStatic }; // static functions are not preserved with `withAITracking`,
+// Hence to use any static function the untransofrmed class has to exported first
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: nit: can we fix the typo in this comment and re-format it so that it's easier to read?

File: RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx [hunk 21]
```diff
@@ -109,3 +111,6 @@ export default class FeedbackItemGroup extends React.Component<IFeedbackItemGrou
     );
   }
 }
+
+
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx]: nit: redundant line

File: RetrospectiveExtension.Frontend/components/selectorCombo.tsx (hunk 15 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/selectorCombo.tsx]: Agree, we should remove line 307 to avoid confusion.

File: RetrospectiveExtension.Frontend/config/environment.tsx.template (hunk 8 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/config/environment.tsx.template:1]: In the docs, we use the following default values:

```json
{
   "CollaborationStateServiceUrl": "https://my-backend-service.com", // change this to the deployed backend service
   "AppInsightsInstrumentKey": "my_instrumentation_key" // put Instrumentation key here
}
```

Can you either update the docs or here for consistency?

File: RetrospectiveExtension.Frontend/index.tsx [hunk 22]
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 import { isHostedAzureDevOps } from './utilities/azureDevOpsContextHelper';
 import { getProjectId } from './utilities/servicesHelper';
 import './css/main.scss';
+import {  appInsights } from './utilities/external/telemetryClient';
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/index.tsx]: nit: There seems to be a redundant space after `{`

File: RetrospectiveExtension.Frontend/index.tsx (hunk 11 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/index.tsx]: nit: missing space after `,`

Comment (phongcao) [reply RetrospectiveExtension.Frontend/index.tsx]: I don't really understand about those properties (`version`, `average`...) so can you please explain more about them in our standup?

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+    config: {
+        instrumentationKey: environment.AppInsightsInstrumentKey,
+        extensions: [reactPlugin],
+        extensionConfig: {
+          [reactPlugin.identifier]: { history: {} }
+        }
+    }
+});
+appInsights.loadAppInsights();
+export { reactPlugin, appInsights };
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:16]: nit: we should have an empty line above

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts [hunk 24]
```diff
@@ -0,0 +1,17 @@
+import { ReactPlugin } from '@microsoft/applicationinsights-react-js';
+import { ApplicationInsights, SeverityLevel } from "@microsoft/applicationinsights-web"
+import environment from '../../config/environment'
+
+
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts]: nit: redundant line here

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
+import { ReactPlugin } from '@microsoft/applicationinsights-react-js';
+import { ApplicationInsights, SeverityLevel } from "@microsoft/applicationinsights-web"
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts]: nit: use single quote

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 23 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:16]: nit: this file should use 2-space indentation like other files for consistency.

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+import environment from '../../config/environment'
+
+
+const reactPlugin = new ReactPlugin();
+const appInsights = new ApplicationInsights({
+    config: {
+        instrumentationKey: environment.AppInsightsInstrumentKey,
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts]: Do you know what happens if the `instrumentationKey` is empty?

**Review by phongcao — CHANGES_REQUESTED** [review]

**Comment by phongcao** [comment]
Thanks for adding this feature. Most of my comments are related to code formatting. However, I do have one question about the event/metric that we're tracking so let's chat about that more in our stand up.

File: RetrospectiveExtension.Frontend/index.tsx (hunk 11 above)
Comment (NarmathaBala) [reply RetrospectiveExtension.Frontend/index.tsx]: Re:tracking versions - It is a great idea to track version information. AppInsights provides an option to set context with addTelemetryInitializers, that way the context of, say a user or version data is tracked consistently throughout. 
Re: tracking startup time - The trackMetric looks like a test code that need to be removed

File: CONTRIBUTING.md (hunk 3 above)
Comment (mhassaninmsft) [reply CONTRIBUTING.md]: changed file name to `vss-extension-dev` singular form

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 4 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx]: good point. I will create a utility class and name `FeedbackColumnHelperClass`

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx (hunk 7 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: done, created `FeedbackColumnHelper` class

**Comment by mhassaninmsft** [comment]
> 

so withAITracking is "with Application Insights Tracking". I guess the authors of the library could have done a better job with cases I guess

File: RetrospectiveExtension.Frontend/index.tsx (hunk 11 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/index.tsx]: Thanks, it makes sense to me that we should remove the trackMetric code. For tracking version, shouldn't we pull the `version` value from some configuration (for example app insights) object? Also, the `Startup` custom event with `version` property seems confusing to me.

**Comment by phongcao** [comment]
@mhassaninmsft @NarmathaBala Sorry if this is something that I missed in our last demo. Can we pull something like the following picture in our environment to see the view counts of individual components?

image

**Comment by mhassaninmsft** [comment]
> @mhassaninmsft @NarmathaBala Sorry if this is something that I missed in our last demo. Can we pull something like the following picture in our environment to see the view counts of individual components?
> 
> image

@phongcao yes that is what we will be adding to the dashboards. If there are any more insights metrics that you think would be helpful, I would love to pair with you on this task if you have some time

File: RetrospectiveExtension.Frontend/.gitignore (hunk 2 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/.gitignore]: will leave everything in the top level .gitignore since many folks left similar comments

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 19 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx:82]: That is a good question. I guess it is because `withAITracking` takes a class as input and returns a value instead of another class. Hence if we remove the `typeof`, the compiler would complain that a value is being used a class. In Typescript there are little difference between an actual class and an object that has the same fields/parameters as defined in a class. Expect in small ways such as typeof. The original function `React.ReactElement<T>` expected either a class or a string,
`interface ReactElement<P = any, T extends string | JSXElementConstructor<any> = string | JSXElementConstructor<any>>`  I am not sure why the function `withAITracking` in the app insights SDK does not return an object that implments `JSXElementConstructor` otherwise the `typeof` expression would have been needless. I can look deeper into this if needed.

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx (hunk 20 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: removed comments since it is not needed any more

File: .gitignore (hunk 16 above)
Comment (mhassaninmsft) [reply .gitignore]: done added templates in the same place for consistency

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 23 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:16]: The `trackEvent` and `trackMetric` events were for local debugging and testing which I forgot to remove. I removed those lines now

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 26 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts]: no crashes happen and the application silently continues. If app insights is required to start the App, we can add a check to make sure that it is not empty and perhaps issue an error or warning if it is.

File: RetrospectiveExtension.Frontend/config/environment.tsx.template (hunk 8 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/config/environment.tsx.template:1]: done

File: .gitignore [hunk 27]
```diff
@@ -260,9 +260,9 @@ paket-files/
 .DS_Store
 
 # dev files
-RetrospectiveExtension.FrontEnd/vss-extension.dev.json
```
Comment (mhassaninmsft) [inline .gitignore:263]: When the folder contains a `.` gitignore does not honor the file name. that is why I created a separate .gitignore file inside `RetrospectiveExtension.FrontEnd`

**Comment by mhassaninmsft** [comment]
addressed the PR reviews feedback. There might have been some formatting and linting issues that I missed. But with Phong's PR for auto Lints this should not be a problem

**Comment by phongcao** [comment]
@mhassaninmsft Can you resolve conflicts and update your branch with `dev`?

File: RetrospectiveExtension.Frontend/config/environment.tsx [hunk 28]
```diff
@@ -1,7 +1,7 @@
 const environment = {
-  CollaborationStateServiceUrl: '<collaboration state service url>',
+  CollaborationStateServiceUrl: 'https://my-backend-service.com', // change this to the deployed backend service
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/config/environment.tsx:2]: Although `enviornment.tsx`  is gitignored, the file is still needed by the code to compile and the CI/CD tests depend on it. So I left the file as is, but it is still gitignored.

**Comment by mhassaninmsft** [comment]
> 

@phongcao done, resolved and rebased with `dev` branch. Thanks!

File: RetrospectiveExtension.Frontend/components/feedbackBoardPreviewEmail.tsx [hunk 29]
```diff
@@ -16,8 +18,7 @@ interface IFeedbackBoardPreviewEmailProps {
   teamId: string;
   onCopy: () => void;
 }
-
-export default class FeedbackBoardPreviewEmail extends React.Component<IFeedbackBoardPreviewEmailProps, IFeedbackBoardPreviewEmailState> {
+class FeedbackBoardPreviewEmail extends React.Component<IFeedbackBoardPreviewEmailProps, IFeedbackBoardPreviewEmailState> {
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackBoardPreviewEmail.tsx:22]: nit: missing an empty line above.

File: RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx (hunk 19 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx:82]: No need, thanks for your explanation!

File: RetrospectiveExtension.Frontend/components/feedbackColumn.tsx [hunk 30]
```diff
@@ -300,3 +258,52 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
...
+      isInteractable: isInteractable,
+      shouldHaveFocus: columnItem.shouldHaveFocus ? true : false,
+      hideFeedbackItems: columnProps.hideFeedbackItems,
+      userIdRef: columnItem.feedbackItem.userIdRef,
+      onVoteCasted: columnProps.onVoteCasted
+    }
+  }
+}
+
+export default  withAITracking(reactPlugin, FeedbackColumn);
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackColumn.tsx]: nit: redundant space after default

File: RetrospectiveExtension.Frontend/components/feedbackColumn.tsx [hunk 31]
```diff
@@ -143,51 +145,7 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
...
-      isInteractable: isInteractable,
-      shouldHaveFocus: columnItem.shouldHaveFocus ? true : false,
-      hideFeedbackItems: columnProps.hideFeedbackItems,
-      userIdRef: columnItem.feedbackItem.userIdRef,
-      onVoteCasted: columnProps.onVoteCasted
-    }
-  }
-
-  private renderFeedbackItems = () => {
+  protected renderFeedbackItems = () => {
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackColumn.tsx]: Seems to me that this doesn't need to be changed to `protected` right?

Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackColumn.tsx]: yeah, changed it back to private. It was some experiments I was doing with the static helper class that I created, but forgot to revert it back to private.

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx [hunk 32]
```diff
@@ -977,3 +963,26 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
     );
   }
 }
+
+export class FeedbackItemHelper
+{
+// Handle linking/grouping workitems and reload any updated items.
+public static handleDropFeedbackItemOnFeedbackItem = async (feedbackItemProps: IFeedbackItemProps, droppedItemId: string, targetItemId: string) => {
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: nit: something wrong with the indentation for this method

File: RetrospectiveExtension.Frontend/index.tsx [hunk 33]
```diff
@@ -6,13 +6,9 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 import { isHostedAzureDevOps } from './utilities/azureDevOpsContextHelper';
 import { getProjectId } from './utilities/servicesHelper';
 import './css/main.scss';
+import { appInsights } from './utilities/external/telemetryClient';
 
 import FeedbackBoardContainer, { FeedbackBoardContainerProps } from './components/feedbackBoardContainer';
```
Comment (phongcao) [inline RetrospectiveExtension.Frontend/index.tsx:11]: Why did you remove `initializeIcons()` method?

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 26 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts]: That's good enough for now, thanks!

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 23 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:16]: Thanks for removing those lines. Can you also help format the file to use 2-space indent if it hasn't already used it?

File: RetrospectiveExtension.Frontend/index.tsx (hunk 33 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/index.tsx:11]: added it back, it was between two lines of TODO and must have been deleted by mistake

Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/index.tsx:11]: good catch! :)

File: RetrospectiveExtension.Frontend/components/feedbackItem.tsx (hunk 32 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/components/feedbackItem.tsx]: fixed indentation

File: RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts (hunk 23 above)
Comment (mhassaninmsft) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:16]: used 2 line indents in file formatting options

**Review by phongcao — APPROVED** [review]

**Comment by phongcao** [comment]
Thanks for addressing all feedback, great work! @mhassaninmsft @NarmathaBala :+1:

**Comment by mhassaninmsft** [comment]
> Thanks for addressing all feedback, great work! @mhassaninmsft @NarmathaBala

Thank you Phong for the quick and thorough review. Really appreciate it
